src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/ArrayOfFrozenObjectsNode.cs

```diff
@@ -1,34 +1,57 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System.Collections.Generic;
+using System;
 
+using Internal.Text;
 using Internal.TypeSystem;
 
 namespace ILCompiler.DependencyAnalysis
 {
-    public class ArrayOfFrozenObjectsNode<TEmbedded> : ArrayOfEmbeddedDataNode<TEmbedded>
-        where TEmbedded : EmbeddedObjectNode
+    public class ArrayOfFrozenObjectsNode : DehydratableObjectNode, ISymbolDefinitionNode
```

Comment:
Dehydratable I assume means serializable. This is an interesting design. What happens if there are cycles in this graph? Are they correctly preserved?

Reply:
Dehydratable means we will run the code from #77884 on this and "dehydrate" pointers and excess zeros in the data blob this node generated into a more succinct representation in the executable image. It will get "rehydrated" at startup.

For cycles in the graph of frozen objects: those can't happen - we don't allow freezing objects with mutable reference-typed fields. The write barriers don't handle frozen segment and the GC wouldn't handle references from those either. The freezing/serialization algorithm wouldn't probably have problems with it.

Reply:
Not having writeable fields does not prevent cycles. The following will have a cycle:

```cs
class c1
{
    private readonly c1 Self;
    c1() => Self = this;
}
```

Can we have this in frozen objects?  

(I am mostly curious if there are other restrictions that will prevent cycles. I do not think dehydration will have problems)

Reply:
That one bails at the stfld:

https://github.com/dotnet/runtime/blob/1076cf52f60a845cf59e316345e2e599a6fcd3ac/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs#L605-L608

In theory the cycle should be fine because the reference would be just a reloc to another symbol and the data blob associated with that symbol has another reloc.

I think the only reference-typed instance fields we currently really allow are those of delegates to instance methods, and those are pretty special.